GVFS/GVFS.Common/LibGit2RepoPool.cs

```diff
@@ -90,6 +140,20 @@ private void ReturnToPool(LibGit2Repo repo)
             }
         }
 
+        private void ResetRepoDisposalTimer()
+        {
+            this.repoDisposalTimer.Change(this.repoDisposalDueTime, this.repoDisposalPeriod);
+        }
+
+        private void TryDropARepo()

```

Comment:
High level question on the approach being used in the PR.

Based on my understanding of the changes it looks like the approach is:

- Periodically dispose `LibGit2Repo` that are not in use
- Assuming there is not disk I/O that requires the use of any repos, eventually the active count will drop to 0
- If the active repo count is 0 when the packfile maintenance job runs it won't run into problems from open libgit2 handles

Is that correct?

If that is the case, what are your thoughts on a slightly different approach?

- Keep all `LibGit2Repo` repos alive until the packfile maintenance job is ready to run
- As a first step of the packfile maintenance job, have it call a `TryDropAllRepos` that does the same thing as `TryDropARepo` except it does it for all the repos in the pool.

A potential benefit to the second approach would be that we avoid the overhead you mentioned in a separate comment:

> The biggest cost comes in the form of startup time when a request come in, as the throughput will not be a huge problem (we don't dispose after every request, just after no requests after a while).
> ... Roughly, this means that every time we have a "pool miss" we will get a 15-16ms delay on each request until the pool is full (assuming that we are receiving a lot of parallel requests, such as during a build). 

If the packfile maintenance job is the only thing that needs the repos to be disposed we can keep all of them hot and ready to go until the the packfile maintenance job needs to run.

If there is a desire to have a minimum 15 second cool-off period before disposing the repos this approach could be adjusted so that `TryDropAllRepos` would check if all of the repos are in the pool and if it's been at least 15 seconds since the last repo was returned to the pool.

Reply:
I'm pretty sure there are side benefits to this approach.

* This will free native memory that is accumulating in the libgit2 repos. We've definitely seen some OOMs in the wild, and this may be part of the cause.

* There is some startup cost to these long-lived repos when we hit an object they don't know about because new packfiles were added by prefetch. The libgit2 code needs to re-scan the pack directory anyway, so that presents a cost not measured by my test.

* The startup cost is limited to a batch of operations that happen within a window. The biggest reason we have a pool is for things like running a build. This may add a small static cost, but it is paid once at the beginning of a build and the continued file access should reduce the cost. If the build is very rarely hydrating a file, then we pay at most 15 ms per 15 minutes.

The points above are to (hopefully) justify clearing the pool regularly without too much of a cost. To add a condition that the packfile maintenance step doesn't run unless the pool is empty is a good idea, but the idea should be reversed: the step should ask if it is already empty (like how we check that there are no running Git processes), but not try to empty the pool itself. By testing if we should run every 6 hours, we will likely find a time where the pool is empty every day to run our step.

Reply:
> - This will free native memory that is accumulating in the libgit2 repos
> - The libgit2 code needs to re-scan the pack directory anyway, so that presents a cost not measured by my test

Both good points 👍 

> If the build is very rarely hydrating a file, then we pay at most 15 ms per 15 minutes.

Did you mean per 15 seconds here? (15s is the period for disposing repos).

One other related question, now that we can fill the pool up on demand, should we still be starting with `size` active repos when `LibGit2RepoPool` is created?  Or should we start with a subset and create more as needed?

Reply:
**Sorry!** You probably didn't see [this comment](https://github.com/Microsoft/VFSForGit/pull/655#discussion_r244025234) about how the current values are super-short on purpose for testing. We should find good values that are less aggressive.

I was thinking a 15 _minute_ due time and a 1 _minute_ period. This means we would not remove a single repo from the pool until 15 minutes after the latest file hydration, and we would only drop to none in the pool after ~16 more minutes (my machine registers as 8 cores, so it loads 16 to the pool).